### PR TITLE
URA-872 - Slack notifications

### PR DESCRIPTION
**Pull request by jethror1**
- add optional Slack notifications on completing / failing upload(s)
- new functions for message formatting and sending post request added in `utils/slack.py`
- added unit tests for new functions
- updated readme

- - -
This change is [Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/20)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,18 +41,21 @@ Available inputs for `monitor`:
...
 The top level keys that may be defined include:
-* `max_cores` (int): maximum number of CPU cores to split uploading across (default is the maximum available)
-* `max_threads` (int): the maximum number of threads to use per CPU core
-* `log_level` (str): the level of logging to set, available options are defined [here](https://docs.python.org/3/library/logging.html#logging-levels)
-* `log_dir` (str): path to where to store logs (default is `/var/log/s3_upload`)
+* `max_cores` (int | optional): maximum number of CPU cores to split uploading across (default: maximum available)
+* `max_threads` (int | optional): the maximum number of threads to use per CPU core
+* `log_level` (str | optional): the level of logging to set, available options are defined [here](https://docs.python.org/3/library/logging.html#logging-levels)
+* `log_dir` (str | optional): path to where to store logs (default: `/var/log/s3_upload`)
+* `slack_log_webhook` (str | optional): Slack webhook URL to use for sending notifications on successful uploads, will try use `slack_alert_webhook` if not specified.
```
Comment (fletchercr) [inline README.md:48]: Why does it use 'slack_alert_webhook' if slack_log_webhook is not specified? And vice versa?

File: s3_upload/s3_upload.py [hunk 2]
```diff
@@ -288,15 +292,56 @@ def monitor_directories_for_upload(config, dry_run):
             failed_files=failed_upload,
         )
 
+        if log_data["uploaded"]:
+            upload_state = "Fully"
+            successfully_uploaded.append(log_data["run_id"])
+        else:
+            upload_state = "Partially"
+            failed_upload.append(log_data["run_id"])
```
Comment (fletchercr) [inline s3_upload/s3_upload.py:300]: Is it fine that the previously empty list variable `failed_upload` that you're appending to here has already been assigned a value from the `multi_core_upload()` function? So we're appending to the output of `multi_core_upload` here?

File: s3_upload/s3_upload.py [hunk 3]
```diff
@@ -288,15 +292,56 @@ def monitor_directories_for_upload(config, dry_run):
             failed_files=failed_upload,
         )
 
+        if log_data["uploaded"]:
```
Comment (fletchercr) [inline s3_upload/s3_upload.py:295]: I may be missing something here or thinking of something eles, but what is `log_data["uploaded"]`? Is it not `log_data["completed"]`?

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/20)* status: 0 of 5 files reviewed, 3 unresolved discussions (waiting on @fletchercr)

___
*[`README.md` line 48 at r1](https://reviewable.io/reviews/eastgenomics/s3_upload/20#gh-1810643551:-O9oRu6C7DEsBizjAYZ1:bz790ds) ([raw file](https://github.com/eastgenomics/s3_upload/blob/d0454f466a72849b223044f6088f28a9a14bd733/README.md#L48)):*
<details><summary><i>Previously, fletchercr wrote…</i></summary><blockquote>

Why does it use 'slack_alert_webhook' if slack_log_webhook is not specified? And vice versa?
</blockquote></details>

it was so if both were to go to the same endpoint that one could just be defined and used for either
___
*[`s3_upload/s3_upload.py` line 295 at r1](https://reviewable.io/reviews/eastgenomics/s3_upload/20#gh-1810677971:-O9oTEHBChN8YNcoSQA_:b65ur3p) ([raw file](https://github.com/eastgenomics/s3_upload/blob/d0454f466a72849b223044f6088f28a9a14bd733/s3_upload/s3_upload.py#L295)):*
<details><summary><i>Previously, fletchercr wrote…</i></summary><blockquote>

I may be missing something here or thinking of something eles, but what is `log_data["uploaded"]`? Is it not `log_data["completed"]`?
</blockquote></details>

nope you're correct, changed it
___
*[`s3_upload/s3_upload.py` line 300 at r1](https://reviewable.io/reviews/eastgenomics/s3_upload/20#gh-1810675562:-O9oTGVp85ye-ISUWiNo:bpsed1j) ([raw file](https://github.com/eastgenomics/s3_upload/blob/d0454f466a72849b223044f6088f28a9a14bd733/s3_upload/s3_upload.py#L300)):*
<details><summary><i>Previously, fletchercr wrote…</i></summary><blockquote>

Is it fine that the previously empty list variable `failed_upload` that you're appending to here has already been assigned a value from the `multi_core_upload()` function? So we're appending to the output of `multi_core_upload` here?
</blockquote></details>

no I reused that variable name when I shouldn't (this is what happens when I try write at the end of day on a Friday), changed both to be prefixed with `runs_` for the uploaded and failed run lists

**Review by fletchercr — APPROVED** [review]

**Pull request merged**